// @sentry/node@8.2.1/esm/integrations/node-fetch.js
import { SpanKind } from '@opentelemetry/api';
import { defineIntegration, addBreadcrumb } from '@sentry/core';
import { addOpenTelemetryInstrumentation, getSpanKind, getRequestSpanData } from '@sentry/opentelemetry';
import { logger } from '@sentry/utils';
import { DEBUG_BUILD } from '../debug-build.js';
import { NODE_MAJOR } from '../nodeVersion.js';
import { addOriginToSpan } from '../utils/addOriginToSpan.js';

const _nativeNodeFetchIntegration = ((options = {}) => {
  const _breadcrumbs = typeof options.breadcrumbs === 'undefined' ? true : options.breadcrumbs;
  const _ignoreOutgoingRequests = options.ignoreOutgoingRequests;

  async function getInstrumentation() {
    // Only add NodeFetch if Node >= 18, as previous versions do not support it
    if (NODE_MAJOR < 18) {
      DEBUG_BUILD && logger.log('NodeFetch is not supported on Node < 18, skipping instrumentation...');
      return;
    }

    try {
      const pkg = await import('opentelemetry-instrumentation-fetch-node');
      return new pkg.FetchInstrumentation({
        ignoreRequestHook: (request) => {
          const url = request.origin;
          return _ignoreOutgoingRequests && url && _ignoreOutgoingRequests(url);
        },
        onRequest: ({ span }) => {
          _updateSpan(span);

          if (_breadcrumbs) {
            _addRequestBreadcrumb(span);
          }
        },
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
      } );
    } catch (error) {
      // Could not load instrumentation
      DEBUG_BUILD && logger.log('Could not load NodeFetch instrumentation.');
    }
  }

  return {
    name: 'NodeFetch',
    setupOnce() {
      // eslint-disable-next-line @typescript-eslint/no-floating-promises
      getInstrumentation().then(instrumentation => {
        if (instrumentation) {
          addOpenTelemetryInstrumentation(instrumentation);
        }
      });
    },
  };
}) ;

const nativeNodeFetchIntegration = defineIntegration(_nativeNodeFetchIntegration);

/** Update the span with data we need. */
function _updateSpan(span) {
  addOriginToSpan(span, 'auto.http.otel.node_fetch');
}

/** Add a breadcrumb for outgoing requests. */
function _addRequestBreadcrumb(span) {
  if (getSpanKind(span) !== SpanKind.CLIENT) {
    return;
  }

  const data = getRequestSpanData(span);
  addBreadcrumb({
    category: 'http',
    data: {
      ...data,
    },
    type: 'http',
  });
}

export { nativeNodeFetchIntegration };
